src/field_theory/separable.lean

```diff
@@ -308,200 +449,100 @@ not_lt_of_le (nat_degree_le_of_dvd this h) $ nat_degree_derivative_lt h⟩
-begin
-  refine f.induction_on' (λ a b ha hb, _) (λ n a, _),
-  { rw [alg_hom.map_add, map_add, ha, hb, add_pow_char], },
-  { rw [expand_monomial, map_monomial, monomial_eq_C_mul_X, monomial_eq_C_mul_X,
-        mul_pow, ← C.map_pow, frobenius_def],
-    ring_exp }
-end
-
-theorem map_expand_pow_char (f : polynomial F) (n : ℕ) :
-   map ((frobenius F p) ^ n) (expand F (p ^ n) f) = f ^ (p ^ n) :=
-begin
-  induction n, { simp [ring_hom.one_def] },
-  symmetry,
-  rw [pow_succ', pow_mul, ← n_ih, ← expand_char, pow_succ, ring_hom.mul_def, ← map_map, mul_comm,
-      expand_mul, ← map_expand (nat.prime.pos hp.1)],
-end
-
-theorem expand_contract {f : polynomial F} (hf : f.derivative = 0) :
-  expand F p (contract p f) = f :=
-begin
-  ext n, rw [coeff_expand hp.1.pos, coeff_contract], split_ifs with h,
-  { rw nat.div_mul_cancel h },
-  { cases n, { exact absurd (dvd_zero p) h },
-    have := coeff_derivative f n, rw [hf, coeff_zero, zero_eq_mul] at this, cases this, { rw this },
-    rw [← nat.cast_succ, char_p.cast_eq_zero_iff F p] at this,
-    exact absurd this h }
-end
-
 theorem separable_or {f : polynomial F} (hf : irreducible f) : f.separable ∨
   ¬f.separable ∧ ∃ g : polynomial F, irreducible g ∧ expand F p g = f :=
-if H : f.derivative = 0 then or.inr
-  ⟨by rw [separable_iff_derivative_ne_zero hf, not_not, H],
-  contract p f,
-  by haveI := is_local_ring_hom_expand F hp.1.pos; exact
-    of_irreducible_map ↑(expand F p) (by rwa ← expand_contract p H at hf),
-  expand_contract p H⟩
+if H : f.derivative = 0 then
+begin
+  unfreezingI { rcases p.eq_zero_or_pos with rfl | hp },
+  { haveI := char_p.char_p_to_char_zero F,
+    have := nat_degree_eq_zero_of_derivative_eq_zero H,
+    have := (nat_degree_pos_iff_degree_pos.mpr $ degree_pos_of_irreducible hf).ne',
+    contradiction },
+  haveI := is_local_ring_hom_expand F hp,
+  exact or.inr
+        ⟨by rw [separable_iff_derivative_ne_zero hf, not_not, H],
+        contract p f,
+        of_irreducible_map ↑(expand F p) (by rwa ← expand_contract p H hp.ne' at hf),
+        expand_contract p H hp.ne'⟩
+end
 else or.inl $ (separable_iff_derivative_ne_zero hf).2 H
 
-theorem exists_separable_of_irreducible {f : polynomial F} (hf : irreducible f) (hf0 : f ≠ 0) :
+theorem exists_separable_of_irreducible {f : polynomial F} (hf : irreducible f) (hp : p ≠ 0) :
   ∃ (n : ℕ) (g : polynomial F), g.separable ∧ expand F (p ^ n) g = f :=
 begin
+  replace hp : p.prime := (char_p.char_is_prime_or_zero F p).resolve_right hp,
   unfreezingI {
     induction hn : f.nat_degree using nat.strong_induction_on with N ih generalizing f },
   rcases separable_or p hf with h | ⟨h1, g, hg, hgf⟩,
   { refine ⟨0, f, h, _⟩, rw [pow_zero, expand_one] },
   { cases N with N,
     { rw [nat_degree_eq_zero_iff_degree_le_zero, degree_le_zero_iff] at hn,
       rw [hn, separable_C, is_unit_iff_ne_zero, not_not] at h1,
+      have hf0 : f ≠ 0 := by unfreezingI { rintro rfl, exact not_irreducible_zero hf },
```

Comment:
Oh, I also used this lemma a fair amount; is this worth having as a lemma? Something like `separable.ne_zero`?